core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -273,7 +273,7 @@ private[spark] class DiskBlockManager(
       Utils.getConfiguredLocalDirs(conf).foreach { rootDir =>
         try {
           val mergeDir = new File(rootDir, mergeDirName)
-          if (!mergeDir.exists()) {
+          if (!mergeDir.exists() || mergeDir.listFiles().length < subDirsPerLocalDir) {
```

Comment:
If we are calling listFiles anyway - why not validate they match the expected pattern ?

Reply:
Sorry, I didn't understand what you said. Could you please describe it again？

Reply:
Instead of only checking for the count, we can check if the files returned by listFiles match the expected filenames.

Reply:
It is very common to use multiple Executors for the same Application on the same machine, and as you say, the `listFiles` operation is almost inevitable, so I think it is necessary to think about what you are saying.
First of all, it would be possible to cache the list of subdirectories in memory after `listFiles`, and then check the existence of the subdirectories when they are created, but this would use some memory (albeit small, and we could manually set it to null after creating the directory)
I'm not sure what your concern is, but I can try to understand it, is it that there are dirty directories under the merge directory? Because that would cause some of the subdirectories not to be created. But I think this worry is not necessary, because under normal circumstances, due to permission related issues, there will be no dirty directories in the merge directory, so there is no problem to determine the number of subdirectories directly.

Reply:
I am not sure I follow the comment above.
What I meant was, instead of simply checking for expected file count - check for the actual directories we expect to find there.

```
  val dirs = Option(mergeDir.listFiles()).map(_.filter(_.isDirectory)).map(_.toSet)
  if (!mergeDir.exists() || dirs.isEmpty || 
        (0 until subDirsPerLocalDir).map(dirNum => new File(mergeDir,  "%02x".format(dirNum))).exists(f => !dirs.get.contains(f)))

```

Reply:
I understand what you mean, but I still don't think it's necessary to check if the subdirectories match, based on the following two points
1. when creating a subdirectory, the file system ensures that it will not be created repeatedly (`if(!subDir.exists())`)
2. theoretically, mergeDir is managed directly by each App itself, and there will be no subdirectories that don't match our expectations
Therefore, determining the number of subdirectories is the same as matching each subdirectory.

Reply:
Having said that, since the directory in this case is entirely managed by ESS currently, we can technically add it later as well - just makes it a bit less robust

Reply:
> 

agree

Reply:
> Having said that, since the directory in this case is entirely managed by ESS currently, we can technically add it later as well - just makes it a bit less robust

ok, tks for your review